app/api/ada/lib/utils.js

```diff
@@ -65,6 +67,48 @@ const _getTxCondition = (state: string): AdaTransactionCondition => {
   return 'CPtxWontApply';
 };
 
+export class CborIndefiniteLengthArray {
+  elements: Array<any>;
+  constructor(elements: Array<any>) {
+    this.elements = elements;
+  }
+  encodeCBOR(encoder: cbor.Encoder) {
+    return encoder.push(
+      Buffer.concat([
+        Buffer.from([0x9f]), // indefinite array prefix
+        ...this.elements.map((e) => cbor.encode(e)),
+        Buffer.from([0xff]), // end of array
+      ])
+    );
+  }
+}
+
+export function rustRawTxToId(rustTxBody: RustRawTxBody): string {
+  if (!rustTxBody) {
+    throw new Error('Cannot decode inputs from undefined transaction!');
+  }
+  try {
+    const [inputs, outputs, attributes] = decodedTxToAux(cbor.decode(Buffer.from(rustTxBody)));
+    const enc = cbor.encode([
+      new CborIndefiniteLengthArray(inputs),
+      new CborIndefiniteLengthArray(outputs),
+      attributes
+    ]);
+    // eslint-disable-next-line
+    return Buffer.from(Blake2b.blake2b_256(enc)).toString('hex');
+  } catch (e) {
+    throw new Error('Failed to convert raw transaction to ID! ' + stringifyError(e));
+  }
+}
+
+function decodedTxToAux(decodedTx) {
```

Comment:
how painful is to add cbor types for this one?

Reply:
Added some type-descriptors, but the point is that this function still has to take `any` as its parameter, and then force-cast the result type based on the array length. There's no better way to check the cbor-type =\